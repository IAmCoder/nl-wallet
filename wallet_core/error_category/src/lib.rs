#[cfg(feature = "sentry")]
pub mod sentry;

use std::{
    fmt::{self, Display, Formatter},
    str::FromStr,
};

pub use error_category_derive::{sentry_capture_error, ErrorCategory};

const CRITICAL: &str = "critical";
const EXPECTED: &str = "expected";
const UNEXPECTED: &str = "unexpected";
const PD: &str = "pd";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Category {
    Expected,     // Don't report to Sentry
    Critical,     // Report Error to Sentry with contents
    PersonalData, // Report Error to Sentry without contents
    Unexpected,   // Should never occur in the Wallet, log error and report to Sentry without contents
}

#[derive(Debug, thiserror::Error)]
#[error("not a category: {0}")]
pub struct UnrecognizedCategory(String);

impl Display for Category {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            Self::Expected => f.write_str(EXPECTED),
            Self::Critical => f.write_str(CRITICAL),
            Self::PersonalData => f.write_str(PD),
            Self::Unexpected => f.write_str(UNEXPECTED),
        }
    }
}

impl FromStr for Category {
    type Err = UnrecognizedCategory;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            EXPECTED => Ok(Self::Expected),
            CRITICAL => Ok(Self::Critical),
            PD => Ok(Self::PersonalData),
            UNEXPECTED => Ok(Self::Unexpected),
            _ => Err(UnrecognizedCategory(s.to_string())),
        }
    }
}

pub trait ErrorCategory {
    fn category(&self) -> Category;
}
