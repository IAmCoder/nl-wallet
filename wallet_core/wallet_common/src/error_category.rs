pub use wallet_common_macros::{sentry_capture_error, ErrorCategory};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Category {
    Expected,     // Don't report to Sentry
    Critical,     // Report Error to Sentry, with contents
    PersonalData, // Report Error to Sentry, without contents
}

pub trait ErrorCategory {
    fn category(&self) -> Category;
}
