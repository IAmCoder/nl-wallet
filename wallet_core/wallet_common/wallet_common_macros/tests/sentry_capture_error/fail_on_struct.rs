use wallet_common::error_category::handle_error_category;

#[handle_error_category]
struct Foo {}

fn main() {}
