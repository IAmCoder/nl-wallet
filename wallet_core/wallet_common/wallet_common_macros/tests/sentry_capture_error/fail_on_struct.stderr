error: attribute macro `handle_error_category` not supported here
 --> tests/handle_error_category/fail_on_struct.rs:4:1
  |
4 | struct Foo {}
  | ^^^^^^
